File: fa_plugin/agents/financial_agent.py
"""Financial Agent - 财务建模与估值分析代理."""

from typing import Any, Dict
import pandas as pd
import numpy as np

from fa_plugin.agents.base import BaseAgent
from fa_plugin.types.models import UserIntent


class FinancialAgent(BaseAgent):
    """财务建模Agent"""

    async def initialize(self, context: Any) -> None:
        """初始化"""
        self.logger.info("[Financial Agent] 初始化完成")
        self.initialized = True

    async def handle(self, intent: UserIntent, context: Any) -> Dict[str, Any]:
        """处理财务相关请求（统一入口）"""
        return await self.build_model(intent, context)

    async def build_model(self, intent: UserIntent, context: Any) -> Dict[str, Any]:
        """
        构建财务模型

        Args:
            intent: 用户意图
            context: FA Context

        Returns:
            响应字典
        """
        try:
            # 获取项目信息
            project_info = intent.entities.get("project_info", {})
            years = context.config.financial_projection_years

            # 构建三表模型
            model = self._build_three_statement_model(project_info, years)

            # 格式化输出
            formatted = self._format_financial_model(model)

            return self._format_response(
                f"💰 **{years}年期财务模型**\n\n{formatted}",
                model=model,
            )

        except Exception as e:
            return self._format_error(e)

    async def analyze_valuation(self, intent: UserIntent, context: Any) -> Dict[str, Any]:
        """
        估值分析

        Args:
            intent: 用户意图
            context: FA Context

        Returns:
            响应字典
        """
        try:
            project_info = intent.entities.get("project_info", {})

            # 执行估值分析
            valuation = self._calculate_valuation(project_info)

            # 格式化输出
            formatted = self._format_valuation(valuation)

            return self._format_response(
                f"🎯 **估值分析报告**\n\n{formatted}",
                valuation=valuation,
            )

        except Exception as e:
            return self._format_error(e)

    def _build_three_statement_model(
        self, project_info: Dict[str, Any], years: int
    ) -> Dict[str, pd.DataFrame]:
        """构建三表模型（使用Pandas）"""
        # 损益表
        income_statement = pd.DataFrame({
            "年份": [f"第{i+1}年" for i in range(years)],
            "营业收入": [100 * (1.5 ** i) for i in range(years)],
            "营业成本": [40 * (1.4 ** i) for i in range(years)],
            "毛利润": [60 * (1.6 ** i) for i in range(years)],
            "运营费用": [30 * (1.3 ** i) for i in range(years)],
            "净利润": [30 * (1.8 ** i) for i in range(years)],
        })

        # 资产负债表（简化）
        balance_sheet = pd.DataFrame({
            "年份": [f"第{i+1}年" for i in range(years)],
            "总资产": [200 * (1.4 ** i) for i in range(years)],
            "总负债": [80 * (1.2 ** i) for i in range(years)],
            "股东权益": [120 * (1.5 ** i) for i in range(years)],
        })

        # 现金流量表（简化）
        cash_flow = pd.DataFrame({
            "年份": [f"第{i+1}年" for i in range(years)],
            "经营活动现金流": [40 * (1.6 ** i) for i in range(years)],
            "投资活动现金流": [-20 * (1.3 ** i) for i in range(years)],
            "筹资活动现金流": [10 * (1.2 ** i) for i in range(years)],
            "现金净增加": [30 * (1.5 ** i) for i in range(years)],
        })

        return {
            "income_statement": income_statement,
            "balance_sheet": balance_sheet,
            "cash_flow": cash_flow,
        }

    def _calculate_valuation(self, project_info: Dict[str, Any]) -> Dict[str, Any]:
        """计算估值"""
        # 简化的估值计算
        return {
            "method": "可比公司法",
            "low": 3500,
            "mid": 5000,
            "high": 6500,
            "comparables": [
                {"name": "可比公司A", "valuation": 4000},
                {"name": "可比公司B", "valuation": 6000},
            ],
            "recommendations": [
                "基于当前市场环境，建议估值区间在3500-6500万元",
                "建议融资时采用中性估值5000万元",
            ],
        }

    def _format_financial_model(self, model: Dict[str, pd.DataFrame]) -> str:
        """格式化财务模型输出"""
        income_df = model["income_statement"]

        result = f"""
📈 **损益表** (单位: 万元)
{income_df.to_string(index=False)}

💼 **资产负债表** (单位: 万元)
{model["balance_sheet"].to_string(index=False)}

💵 **现金流量表** (单位: 万元)
{model["cash_flow"].to_string(index=False)}

📊 **关键指标**
- 年复合增长率(CAGR): 50%
- 毛利率: 60%
- 净利率: 30%
"""
        return result

    def _format_valuation(self, valuation: Dict[str, Any]) -> str:
        """格式化估值输出"""
        return f"""
**估值方法**: {valuation['method']}

**估值区间** (单位: 万元)
- 保守估计: {valuation['low']:,.0f}
- 中性估计: {valuation['mid']:,.0f}
- 乐观估计: {valuation['high']:,.0f}

**可比公司**
{chr(10).join([f"- {c['name']}: {c['valuation']:,.0f}万元" for c in valuation.get('comparables', [])])}

**建议**
{chr(10).join([f"• {r}" for r in valuation.get('recommendations', [])])}
"""

File: src/agents/financial-agent.ts
/**
 * Financial Agent - 财务建模与估值分析代理
 */

import type {
  SkillContext,
  SkillResponse,
  UserIntent,
  FinancialProjection,
  ValuationAnalysis,
  ProjectInfo,
} from '../types.js';
import { buildFinancialModel } from '../tools/financial-model.js';
import { calculateValuation } from '../tools/valuation.js';

export class FinancialAgent {
  async initialize(context: SkillContext): Promise<void> {
    console.log('[Financial Agent] 初始化完成');
  }

  /**
   * 构建财务模型
   */
  async buildModel(
    intent: UserIntent | Record<string, any>,
    context: SkillContext
  ): Promise<SkillResponse> {
    try {
      const projectInfo = intent.entities?.projectInfo || intent;

      // 构建三表模型
      const model = await buildFinancialModel(projectInfo, context);

      return {
        type: 'text',
        content: this.formatFinancialModel(model),
        metadata: { model },
      };
    } catch (error) {
      return {
        type: 'error',
        content: `构建财务模型时出错: ${error instanceof Error ? error.message : String(error)}`,
      };
    }
  }

  /**
   * 估值分析
   */
  async analyzeValuation(
    intent: UserIntent | Record<string, any>,
    context: SkillContext
  ): Promise<SkillResponse> {
    try {
      const projectInfo = intent.entities?.projectInfo || intent;

      // 执行估值分析
      const valuation = await calculateValuation(projectInfo, context);

      return {
        type: 'text',
        content: this.formatValuation(valuation),
        metadata: { valuation },
      };
    } catch (error) {
      return {
        type: 'error',
        content: `估值分析时出错: ${error instanceof Error ? error.message : String(error)}`,
      };
    }
  }

  /**
   * 格式化财务模型输出
   */
  private formatFinancialModel(model: any): string {
    const { income_statement, balance_sheet, cash_flow } = model;

    return `
💰 **三年期财务模型**

📈 **损益表 (单位: 万元)**
${this.formatIncomeStatement(income_statement)}

💼 **资产负债表 (单位: 万元)**
${this.formatBalanceSheet(balance_sheet)}

💵 **现金流量表 (单位: 万元)**
${this.formatCashFlow(cash_flow)}

📊 **关键指标**
- 毛利率: ${model.metrics.gross_margin}%
- 净利率: ${model.metrics.net_margin}%
- 收入增长率: ${model.metrics.revenue_growth}%
- 预计盈亏平衡点: 第 ${model.metrics.break_even_month} 个月

💡 **财务建议**
${model.recommendations.map((r: string) => `• ${r}`).join('\n')}
`;
  }

  private formatIncomeStatement(is: any[]): string {
    const headers = '|  项目  | 第1年 | 第2年 | 第3年 |';
    const separator = '|--------|-------|-------|-------|';
    const rows = [
      `| 营业收入 | ${is[0].revenue} | ${is[1].revenue} | ${is[2].revenue} |`,
      `| 营业成本 | ${is[0].cogs} | ${is[1].cogs} | ${is[2].cogs} |`,
      `| 毛利润 | ${is[0].gross_profit} | ${is[1].gross_profit} | ${is[2].gross_profit} |`,
      `| 销售费用 | ${is[0].sales_expense} | ${is[1].sales_expense} | ${is[2].sales_expense} |`,
      `| 管理费用 | ${is[0].admin_expense} | ${is[1].admin_expense} | ${is[2].admin_expense} |`,
      `| 研发费用 | ${is[0].rd_expense} | ${is[1].rd_expense} | ${is[2].rd_expense} |`,
      `| 营业利润 | ${is[0].operating_profit} | ${is[1].operating_profit} | ${is[2].operating_profit} |`,
      `| 净利润 | ${is[0].net_profit} | ${is[1].net_profit} | ${is[2].net_profit} |`,
    ];

    return [headers, separator, ...rows].join('\n');
  }

  private formatBalanceSheet(bs: any[]): string {
    return `
**资产**
- 流动资产: ${bs[0].current_assets} → ${bs[2].current_assets}
- 固定资产: ${bs[0].fixed_assets} → ${bs[2].fixed_assets}

**负债**
- 流动负债: ${bs[0].current_liabilities} → ${bs[2].current_liabilities}
- 长期负债: ${bs[0].long_term_liabilities} → ${bs[2].long_term_liabilities}

**所有者权益**
- 实收资本: ${bs[0].equity} → ${bs[2].equity}
- 留存收益: ${bs[0].retained_earnings} → ${bs[2].retained_earnings}
`;
  }

  private formatCashFlow(cf: any[]): string {
    return `
- 经营活动现金流: ${cf[0].operating} → ${cf[2].operating}
- 投资活动现金流: ${cf[0].investing} → ${cf[2].investing}
- 筹资活动现金流: ${cf[0].financing} → ${cf[2].financing}
- 期末现金余额: ${cf[0].ending_cash} → ${cf[2].ending_cash}
`;
  }

  /**
   * 格式化估值分析输出
   */
  private formatValuation(valuation: ValuationAnalysis): string {
    const { method, estimated_value, comparables } = valuation;

    let content = `
🎯 **估值分析报告**

**估值方法**: ${this.getMethodName(method)}

**估值区间** (单位: 万元)
- 保守估计: ${estimated_value.low.toLocaleString()}
- 中性估计: ${estimated_value.mid.toLocaleString()}
- 乐观估计: ${estimated_value.high.toLocaleString()}
`;

    if (comparables && comparables.length > 0) {
      content += `\n**可比公司**\n`;
      comparables.forEach((comp) => {
        content += `\n${comp.name}\n`;
        content += `  - 轮次: ${comp.latest_round}\n`;
        content += `  - 估值: ${comp.valuation.toLocaleString()} 万元\n`;
        if (comp.revenue) {
          content += `  - PS 倍数: ${comp.multiple}x\n`;
        }
      });
    }

    content += `\n💡 **估值建议**\n`;
    content += valuation.recommendations.map((r) => `• ${r}`).join('\n');

    return content;
  }

  private getMethodName(method: string): string {
    const names: Record<string, string> = {
      comparable: '可比公司法',
      dcf: '现金流折现法 (DCF)',
      venture_capital: '风险投资法',
      market_multiple: '市场倍数法',
    };
    return names[method] || method;
  }
}
